### ISSUE TITLE
For  Horizontal pod autoscaling enable flag to don't count in currentReplicas pods that are  not Ready

### ISSUE THREAD

**Issue by lumeche** [kubernetes/kubernetes#59975]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
> /kind bug

/kind feature

**What happened**:
We are  starting  to use the HPA (horizontal pod  autoscaling) for one of  our components. What is  special about this  particular  component is  that the warm up period is very long (sometimes  even more  than 30 minutes). During that time, the pod is in Running state but its containers are in Not Ready state.
This  pod  is of course not  able to receive any traffic so the  metric we  use in the HPA is ignoring those  pods that are not  Ready.

The  equation used  by HPA  to calculate the  new  number  of replicas is as  follows
`new replica count=(current replica  count)*(current metric value)/(expected metric)`

Our problem is  that the equation used  by HPA to calculate the  new number  of  replicas  is including in the "current  replica count" the pods that  are  "Not  Ready" so the value  for "new replica count" is not correct. Since this warm up is so long  we end up by having a lot of more instances of our services of what it should  be necessary.  They eventually get deleted once  gradually the metrics start going down but  that will  take a lot of time and a big waste of resources without mentioning  that with a  changing traffic this HPA will never stabilize  

**What you expected to happen**:
I would like to have a flag somewhere  where  I tell HPA to don't include  in the  "current replica count" the pods  that are  in "Not Ready" state

**How to reproduce it (as minimally and precisely as possible)**:
Have a component that takes a lot of time to start and being  able to receive traffic. Then define  a metric that  will ignore  the pods  that  are  Not Ready. Finally create  a  metric that  will scale that component based on  the mentioned  metric

**Anything else we need to know?**:

**Environment**:
- Kubernetes version 1.7
- Cloud provider or hardware configuration:
Azure

**Comment by mattjmcnaughton** [comment]
Hi @lumeche - thanks for reaching out about this!

It looks like this section of the `replica_calculator` [code](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podautoscaler/replica_calculator.go#L128), handles when pods aren't ready. @DirectXMan12 I think you added this behavior in [this commit](https://github.com/kubernetes/kubernetes/commit/2c66d47786465b132f9b5d1d731c4eb9f6c9d686), so I'm interested in your thoughts on whats best here.

If we do end up adding a flag for this, I'd be happy to write the code :)

Also, @lumeche  purely for my own curiosity, I'm curious what type of pod is taking 30 minutes to become ready. If your traffic pattern varies at a rate faster than 30 minutes, it may be difficult for auto-scaling to be helpful.

**Comment by bskiba** [comment]
@lumeche can you also share a bit more about your HPA? What kind of metric type is it? Based on the description I'm suspecting that it is an Object metric, am I right?

**Comment by lumeche** [comment]
So answer @mattjmcnaughton about why our process takes that much time to start. Basically this is a server that in order to do what he needs to do, he needs to load a lot of data in memory (several GBs) and then perform some warm up that are pretty intensive. The idea is that he doesn't need to grab (almost) anything from outside to attend a request so it's latency is extremely low (in the order of millisecond)

By the way. Because of this we have another problem which is that there is not an specific **flag per HPA object** that we could tell k8s to don't try to scale up again passed some time so the new pods have enough time to get ready before the next scale happens. I know that there is a cluster wide flag but we will need to have a different one by resource that we scale. 

Now to answer @bskiba. You pointed another interesting topic. The metric we use is % Utilization. We are able to calculate the level of utilization of our pods and we want to keep this metric in an specific value.

**Comment by eghobo** [comment]
@mattjmcnaughton, @DirectXMan12 do you think we can add this change to 1.10? it's critical bug/feature for us and we are happy to help with code/testing.

**Comment by DirectXMan12** [comment]
It's *really* late for 1.10, especially since we don't have any code here, and I'd like to know more about what's going on here, and especially see if it reproduces on Kubernetes 1.9 (we've had HPA controller bugs in the past).

FWIW, currently unready pods are treated as consuming 0% of their utilization on scale up, and it was intended that this should prevent such issues.  That is, the formula is actually:

```
metrics = readyPodMetrics
usageRatio = avg(metrics)/desiredMetric
if usageRatio > 0 && len(unreadyPods) > 0 {
  metrics = append(metrics, 0, 0, 0, ...)
  usageRatio = avg(metrics)/desiredMetric
}
newReplicas = usageRatio * len(metrics)
```

If the pods are marked as not ready, I find it strange that they would be affecting the scaling that much, unless your other pods are under real heavy load (i.e. I would assume that the unready pods would skew the scaling back down so that the number of pods just stayed constant).

**Comment by eghobo** [comment]
@DirectXMan12 I believe there is different code branch here. 

metric definition
`{
	"type": "Object",
	"object": {
		"target": {
			"kind": "Service",
			"name": "prometheus-exporter"
		},
		"metricName": "PercentUtilization",
		"targetValue": "20"
	}
}`

which should handle as [ObjectMetricSourceType](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podautoscaler/horizontal.go#L226) and [GetObjectMetricReplicas](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podautoscaler/replica_calculator.go#L275) doesn't check Pod state.

**Comment by MaciekPytel** [comment]
I think proper handling of unready pods is a difficult problem that will require significant work. I suspect we may find out we need to start distinguishing pods that are booting up from pods that are 'broken' and it took us quite a while to solve analogous problem with nodes in Cluster Autoscaler (to be clear: I don't know if we will have to do that, that's just my intuition for now).

That being said I agree the formula in GetObjectMetricReplicas will generally lead to scaling-up too much if the pod takes longer to initialize then backoff period in HPA (3 minutes).

Assume it takes pod 4 minutes to boot up and we're currently running 2 pods. Metric value is 20 and target is 10. In this case:
1. HPA scales-up to 4 pods (20/10 utilization * 2 pods).
2. Pods are starting up, they're not doing anything yet.
3. 3 minutes later HPA scales-up to 8 pods (20/10 utilization * 4 pods).

If we only counted ready pods the scale-up in 3 would not occur, because the desired replicas would still be 4 (20/10 * 2). My intuition would be that this is the expected behavior of HPA.

Also if there is no space in the cluster we will just keep scaling-up until we hit max replicas.

The only case where current behavior would be more correct than filtering out unready pods in this formula is if someone uses a metric that is calculated based on total number of replicas. Do we advertise that as the right way to use HPA? If we add `targetAverageValue` to Object metrics in 1.11 this usage pattern will stop making sense.

To sum up - I think we will need to change the formula to only count ready pods at some point. I can see how current behavior could be a problem in any use-case except the one described in paragraph above. I'm inclined to agree with @eghobo and provide some way to change the behavior in 1.10 (a flag or an annotation?), treating it as a bug fix. WDYT @DirectXMan12?

**Comment by DirectXMan12** [comment]
Ah, yeah, I missed that it was object metrics before.  I'm not entirely convinced we need a flag here -- it might just make sense to bring Object metrics in line with all the other metrics types.

On the other hand, there's a subtlety here -- the other metric types don't quite *ignore* unready pods -- they just ignore the metrics from unready pods, which causes a heavy bias in the direction of not scaling, but an overwhelming spike can still cause scaling.  That behavior is not as achievable using object metrics.  I'm inclined to just ignore that subtlety, though, and say that object metrics should ignore unready pods entirely.

**Comment by MaciekPytel** [comment]
I think this is what we should do long term. I was worried about breaking someone who relies on targetValue + averaged metric to get effect similar to targetAverageValue and though about providing something for transitional period, but if you're ok with it I'd rather just change it as you say.

**Comment by DirectXMan12** [comment]
I think it's safe to change as-is -- that feature's still in beta, and it's not an API change, so it shouldn't just hard-break anyone.  Plus, we've change stuff like this before (when we did the initial resource metrics change).

**Comment by lumeche** [comment]
As @eghobo said, we are more than ready to help with  the fix/test if required. This  since this is a major  blocking for us  in  the implementation  of  the HPA in our side

**Comment by mattjmcnaughton** [comment]
@DirectXMan12 @MaciekPytel is anyone actively working on implementing this? If not, I'm happy to. I think I have a fairly clear picture of the change we need to make and will likely have time early this week to get it done.

**Comment by mattjmcnaughton** [comment]
(Also, from the conversation above, to we want to tag this as a release blocker for 1.10)?

**Comment by MaciekPytel** [comment]
@mattjmcnaughton I probably won't be able to work on this before the end of this week. Happy to do a review if you jump in and implement it.

**Comment by DirectXMan12** [comment]
I don't think this qualifies as a release blocking bug (but I'm willing to have someone try and convince me otherwise ;-)).  There doesn't seem to be any behavior here that's explicitly *broken* -- it's just sub-optimal in some use cases.  I'd definitely like to address this in 1.11, though.

**Comment by eghobo** [comment]
it's not release blocker, but to share our experience ;). without this feature we cannot use HPA as is and we need to figure out how to build/run fork (which can be challenging because everything inside controller-manager). also we are planning to upgrade to 1.10 quickly, but not sure for 1.11 (e.g. we are two versions behind now).

**Comment by mattjmcnaughton** [comment]
Sounds like a plan :) Will try and get this done tomorrow morning!

**Comment by mattjmcnaughton** [comment]
@eghobo @lumeche - I submitted a pr for this - would love your eyes on it so you can see if you think it meets your use case :)

**Comment by lumeche** [comment]
I checked the code and it seems good to me. About the comment @MaciekPytel about scaling down if the readyPods<CurrentPods, we should be fine since our metric is already not including the pods that are not ready.

**Comment by lumeche** [comment]
Is there any advance in this ticket? I'm wondering if this made it to k8s 1.10 or if we need to wait for k8s 1.11?

**Comment by MaciekPytel** [comment]
The PR is lgtm-ed and it will merge once the code freeze is lifted. Meaning it will go to 1.11. To get it into 1.10 you would need to get approval for milestone from @DirectXMan12 (sig-autoscaling lead).
